ale_linters/go/gobuild.vim

```diff
@@ -1,17 +1,45 @@
-" Author: dzhou121 <dzhou121@gmail.com>
+" Author: dzhou121 <dzhou121@gmail.com>, Ryan Norris <rynorris@gmail.com>
 " Description: go build for Go files
 
-function! s:FindGobuildScript() abort
-    return g:ale#util#stdin_wrapper . ' .go go build -o /dev/null'
-endfunction
+function! ale_linters#go#gobuild#GetCommand(buffer) abort
+    " Get absolute path to the directory containing the current file.
+    " This directory by definition contains all the files for this go package.
+    let l:this_package = expand('%:p:h')
+
+    " Get a listing of all go files in the directory.
+    " TODO: Handle packages that contain c files.
+    let l:all_files = globpath(l:this_package, '*.go', 1, 1)
+
+    " Filter out the current file since we don't want to include it twice.
+    " We'll then pass this list to go compile and stdin_wrapper will add on
+    " the temporary version of the current file.
+    let l:this_file = expand('%:p:t')
+    call filter(l:all_files, 'fnamemodify(v:val, '':t'') != l:this_file')
+
+    " We're also going to need the system information in order to find the
+    " correct import directory.  We'll pull these from go itself since then
+    " we're guaranteed they'll match what it's looking for.
+    " The version string is of the form 'go version go1.6.2 darwin/amd64'.
+    let l:go_version_string = system('go version')
+    let l:version_part = substitute(l:go_version_string, '.* ', '', '')
+    let l:goos = substitute(l:version_part, '/.*', '', '')
+    let l:goarch = substitute(l:version_part, '.*/', '', '')
 
-let g:ale#util#gobuild_script =
-\   get(g:, 'ale_go_gobuild_script', s:FindGobuildScript())
+    " goarch comes with a troublesome trailing newline,  so strip that off.
+    let l:goarch = substitute(l:goarch, '\n\+$', '', '')
+
+    " Finally build the import path.
+    " TODO: Handle GOPATHs that contain multiple directories.
+    let l:gopath = $GOPATH
+    let l:import_path = l:gopath . '/pkg/' . l:goos . '_' . l:goarch
```

Comment:
Yep.  I've gone ahead and implemented this (but using map/join).
The slightly annoying thing is handling how GOPATH is handled differently on different OSes.